Sources/TensorFlow/Loss.swift

```diff
@@ -287,8 +287,6 @@ public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
 ) -> Tensor<Scalar> {
     // This numerically stable implementation is based on the TensorFlow Python API.
     let maxLogitsWithZero = max(logits, Tensor(0))
-    // Note: `result` is split into two lines to avoid the "compiler is unable to type-check this
-    // expression in reasonable time" error.
-    let result = log(1 + exp(-abs(logits)))
-    return reduction(maxLogitsWithZero - logits * labels + result)
+    let negAbsLogits = max(logits, -logits) // Custom `abs` to compute gradients at `0`.
```

Comment:
Could you please clarify why `max(logits, -logits)` works while `abs(logits)` does not? Is the derivative of `abs` inconsistent with Python TensorFlow?

The old code seems to have been derived directly from the [doc comment for `tf.losses.sigmoid_cross_entropy_with_logits`](https://github.com/tensorflow/tensorflow/blob/00fad90125b18b80fe054de1055770cfb8fe4ba3/tensorflow/python/ops/nn_impl.py#L136):
```
  Hence, to ensure stability and avoid overflow, the implementation uses this
  equivalent formulation
      max(x, 0) - x * z + log(1 + exp(-abs(x)))
```

Reply:
Well both work. It's just a matter of how we approach the evaluation of gradients on `0` entries of the `logits`. Recall that `abs` and `max` are both non-differentiable at that point. Now, in our case there happens to be a discrepancy in the gradient responses of `abs` and `max(x, -x)` for reasons I've yet to see. It could just be how `max` breaks ties. Note that this response is also happening in Python TF (and PyTorch if that matters). 
I chose to use `max(logits, -logits)` because it matches the Python TF implementation of `tf.losses.sigmoidCrossEntropy`. Using `abs` works just as well but the gradients through any `0` entries of `logits` are going to be different. If you'd like, I'd gladly change over to using `abs` as it seems to keep more `0` entries at `0` which could, ever so slightly, help loss convergence.

On another note, I ran a few gradient tests on `abs` and did not find inconsistencies between the Python TF and swift-api versions. All is well on that end.

Reply:
I believe I know what the reason for this is. In this PR, both min and max gradients check for `<=` and `>=` when computing gradients, which means that at 0 they set the gradient to `1`. `abs` uses `sign` which sets it to  0. From a subgradient optimization point of view, I believe `0` will behave better and I would thus rather use `abs`. We should probably change the gradients of `min` and `max` to do the same though, which can be easily done by changing `.<=` and `.>=` to `.>` and `.<` when calling `_vjpMinMaxHelper`. What do you think?

Reply:
@eaplatanios That's exactly it! I also prefer the behavior of `abs` but wasn't sure about breaking parity with the TF implementation. 
I'm +1 on taking out the lower and upper bounds in the grad helper. I'll make the above changes and test things out. Thanks for shedding some light on this. :).

Reply:
That's great, thanks! :) Regarding breaking parity with the TF implementation, I believe that consistency in the behavior of `abs(x)` and `max(x, -x)` is more important and would thus also lean towards ignoring parity with the Python API in this case. @dan-zheng @rxwei what do you guys think?

Reply:
> Regarding breaking parity with the TF implementation, I believe that consistency in the behavior of abs(x) and max(x, -x) is more important and would thus also lean towards ignoring parity with the Python API in this case.

Unfortunately, I believe maintaining parity with Python TensorFlow gradient calculations is important. We want to the ability to check that gradient calculations exactly match Python TensorFlow.

Reply:
Let's just go with the Python version since all implementations are technically correct. The inconsistency pointed out should not really matter in practice. However, please check my other response for some clarification as to which one is the correct Python implementation. :)